pertpy/tools/_distance_tests.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import annotations
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from rich.progress import track
+from sklearn.metrics import pairwise_distances
+from statsmodels.stats.multitest import multipletests
+
+from ._distances import Distance
+
+"""
+DEV NOTES:
+- There will be a single PermutationTest class that internally does one of the following strategies:
+    - recomputes the pairwise distance matrix every time and works for
+      any metric that has .
+    - uses the precomputed pairwise distance matrix and works only for
+    metrics that can be called on the pairwise distance matrix. Will be faster.
+
+TODO:
+- Make PermutationTest allow for multiple controls (accept list of controls)
+- Check if PermutationTest works for all metrics
+"""
+
+
+class PermutationTest:
+    """Runs permutation tests for all groups of cells against a contrast group ("control").
+
+    Args:
+    metric (str): Distance metric to use.
+    n_perms (int): Number of permutations to run.
+    obsm_key (str): Name of embedding to use for distance computation.
+    alpha (float): Significance level.
+    correction (str): Multiple testing correction method.
+    """
+
+    def __init__(
+        self,
+        metric: str,
+        n_perms: int = 1000,
+        obsm_key: str = "X_pca",
+        alpha: float = 0.05,
+        correction: str = "holm-sidak",
+    ):
+        self.metric = metric
+        self.n_perms = n_perms
+        self.obsm_key = obsm_key
+        self.alpha = alpha
+        self.correction = correction
+
+    def __call__(self, adata: AnnData, groupby: str, contrast: str, verbose: bool = True) -> pd.DataFrame:
+        """Run permutation test.
+
+        Args:
+        adata (anndata.AnnData): Annotated data matrix.
+        groupby (str): Key in adata.obs for grouping cells.
+        contrast (str): Name of the contrast group.
+        verbose (bool): Whether to print progress.
+
+        Returns:
+        pandas.DataFrame: Results of the permutation test, with columns:
+            - distance: distance between the contrast group and the group
+            - pvalue: p-value of the permutation test
+            - significant: whether the group is significantly different from the contrast group
+            - pvalue_adj: p-value after multiple testing correction
+            - significant_adj: whether the group is significantly different from the contrast group after multiple testing correction
+        """
+
+        if Distance(self.metric, self.obsm_key).metric_fct.accepts_precomputed:
+            # Much faster if the metric can be called on the precomputed
+            # distance matrix, but not all metrics can do that.
+            return self.test_precomputed(adata, groupby, contrast, verbose)
+        else:
+            return self.test_xy(adata, groupby, contrast, verbose)
+
+    def test_xy(self, adata: AnnData, groupby: str, contrast: str, verbose: bool = True) -> pd.DataFrame:
```

Comment:
Not sure whether the function name is self descriptive TBH.

Could this please get a full docstring with all parameters + returns?